core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1985,12 +1984,15 @@ public void deleteInternal(RpcContext rpcContext, LockedInodePath inodePath,
               job.setCancelState(PersistJob.CancelState.TO_BE_CANCELED);
             }
           }
-          revisedInodesToDelete.add(new Pair<>(alluxioUriToDelete, inodePairToDelete.getSecond()));
         } else {
           unsafeInodes.add(inodeToDelete.getId());
           // Propagate 'unsafe-ness' to parent as one of its descendants can't be deleted
           unsafeInodes.add(inodeToDelete.getParentId());
           failedUris.add(new Pair<>(alluxioUriToDelete.toString(), failureReason));
+
+          // Something went wrong with this path so it cannot be removed normally
+          // Remove the path from further processing
+          inodesToDelete.set(i, null);
```

Comment:
I don't use iterator.remove() because the ArrayList.remove() will incur array copy.

Interestingly, ArrayList.remove() does not copy the whole array, but only left-shifts the elements on the right of the index. So the cost of remove is not strictly `O(n)`.

Setting to null only helps GC on the `Pair` object, but not the underlying locks and `LockedInodePath` so the benefit is just some Bytes. I figure unlocking those paths prematurely makes the code much messier so I'm not doing that.